transactions/read-only/parallel.md

```diff
@@ -0,0 +1,214 @@
+
+```mermaid
+flowchart TD
+    A(((write window))) -->|timer progressing| B[write_window_deadline passed?]
+    B -->|yes| C[read-only trx queue empty?]
+    B -->|no| A
+    C -->|yes| A
+    C -->|no| R(((read window)))
+    
+    R -->|read-only trx executed| S[read-only trx queue empty agreed by all the threads?]
+    S -->|yes| A
+    S -->|no| R
+    R -->|timer progressing| T[read_window_deadline passed?]
+    T -->|yes| A
+    T -->|no| R  
+```
+
+### Requests Starvation Discussion
+- Comparing with executing all requests in the single main thread, offloading read-only transactions to a separate thread pool makes the time used to execute them available for other operations, resulting more operations are processed. 
+- We are most concerned about starvation of NET `signed_block` message. Its priority is medium, which is higher than the majority of read operations. When `read window` is switched to `write window`, if the front operation in the write window queue is block sync, it is likely to be selected for execution. The worst case is a `signed_block` at the front of the queue right after the execution window is switched to `read window`, resulting in the `signed_block` message waits for the whole duration of `read window`. Configuring a small `read-only-read-window-ms` would reduce the waiting time for `signed_block`. Furthermore, `write window` is switched to `read window` only when there are read-only transactions in the read-only transaction queue; this means the main thread has already saved the time required to executing those read-only transactions. The `signed_block` is processed earlier than it would be in the existing single thread approach otherwise.
+- Other operations not safe to read-only transactions are `compute_transaction` (low priority), `push_transaction` (low), `push_transactions` (low), `send_transaction` (low), `send_transaction2` (low), `push_block` (medium_low), `set_whitelist_blacklist` (medium_high), `schedule_protocol_feature_activations` (medium_high), `packed_transaction` (low). Those operations with low priority would not be scheduled even allowed to, if operations safe to read-only transactions are queued. `set_whitelist_blacklist` and `schedule_protocol_feature_activations` are not used frequently, `send_block` mostly not used (can be deprecated).
+- In `read window`, whenever read-only transaction queue becomes empty, `read window` is switched to `write window`. This helps to process any write operations earlier if they are available.
+- To prevent read-only transactions from stuck in `read window` and prevent `write operations` from stuck in `write window` for too long, `read-only-write-window-ms` and `read-only-read-window-ms` should be configured to be small numbers, especially when `read-only-transaction-threads` is set to a big number (which results in more read-only transactions can be executed in shorter time). This promotes rapid toggling of the two windows and reducing starvation.
```

Comment:
> To prevent read-only transactions from getting stuck in the `read-only transaction queue` for too long and to prevent `write operations` from getting stuck in `write operation queue` for too long, `read-only-read-window-time-us` and `read-only-write-window-us` should be configured to be reasonably small numbers. This promotes rapid toggling between the two windows and reduces delays and starvation. However, `read-only-read-window-time-us` should not be too low to give enough time for a reasonable read-only transaction to successfully execute. Additionally, the ratio between `read-only-read-window-time-us` and `read-only-write-window-time-us` should be set to an appropriate amount to avoid starvation of either type of operations; when `read-only-transaction-threads` is set to a large number (and there are enough cores available on the machine to actually justify that number), that ratio can be made smaller.

Reply:
Thanks.